GSA-TTS/FAC
# DRAFT - WIP - POM for Search Page + Cypress tests

This PR is linked to #3485 
This PR will close #3590 
This PR will close #3591 

In this PR:

- I added a plugin `downloadFile` for cypress and added it as a task in `cypress.config.js`
- Under `support` in Cypress directory, in the `commands.js` file the line `require(cypress-downloadfile)` was added.
- `Cypress-downloadfile` was added to `package-lock.json` & `package.json`
- Added `search-tests.cy.js` to e2e, which has two tests 1) basic search test & 2) advance search test
- Added data for search page to fixtures under `search_page_data` this holds all data for search parameters in one single json file vs before where I originally separated each field into it's own separate json files.
- I created a `page_objects` folder w/ file `search-page.js` this holds all web elements for search page + summary report and method/functions for interacting with web element.
- POM for `gen-info `added to `Page Objects` + deleted old `gen-info` cypress test.
- I modified the `login-gov.js` to include an `if else` statement due to the fact that login.gov started to ask if you wanted to add your PIV card as an authentication, I added the `if else` to skip this step.
- I modified the `initialize-audit.js` to import from the `page_objects` folder.
- I added `check-access.js` to the `page_objects` folder and deleted it from the `support` folder.
- I also added some new dependencies to both the `package-lock.json` & `package.json`.

# How to test:

1. Pull this branch, now since we're using a plugin for cypress, we need to make sure to download it locally on our machine or the tests won't work for search. Make sure you're on the `backend` directory of the project and run `npm i --save-dev cypress-downloadfile`. Once installed, you should see in your IDE a new folder titled `downloads` appear in the cypress directory, it should appear just under cypress but before e2e. 

2. Run docker as you normally would and once docker is running on another tab run this command `docker compose exec web /bin/bash` this will get you into docker shell. After that, run this command `python manage.py materialized_views --create` .  (I had to do this since I didn't have materialized view and my search tests would fail but after running command they passed)

3. Now in your terminal, enter your cypress env variables and run npx cypress open 

4. Once cypress is open, run the `full-submission.cy.js` tests should pass may take one or two tries for them but once they are completed run the other test for `search-tests.cy.js` -- Side 
Note: BEFORE running search tests, you'll notice that in your cypress downloads folder you won't have anything in there but AFTER running search tests, a `sf-sac-summary-report.xlsx` and/or a `single-audit-report.pdf` will show up, indicating that the download did work.
# Before Running Search Tests:
<img width="275" alt="Screenshot 2024-04-04 at 12 09 16 PM" src="https://github.com/GSA-TTS/FAC/assets/99770283/893e194b-d16d-49b3-aae9-784b73e6aa0f">


5. Now once you run `search-tests.cy.js` you should able to run a basic search that populates results, selecting a an audit from the table and viewing your summary report and downloading each document successfully.

https://github.com/GSA-TTS/FAC/assets/99770283/65d9f0c9-5872-4efa-b386-15805ab5e3d2



# After Running Search Tests:
<img width="420" alt="Screenshot 2024-04-04 at 12 30 02 PM" src="https://github.com/GSA-TTS/FAC/assets/99770283/de19d2c1-f9a7-4e0e-9b6f-668dbd0d05ce">

6. The same applies for testing `advance search fields` but these are commented out for now since basic search returns results.

7. You can also verify search results with dbeaver by checking in `Materialized Views`:
<img width="1403" alt="Screenshot 2024-04-04 at 12 36 35 PM" src="https://github.com/GSA-TTS/FAC/assets/99770283/afe63595-3420-4f37-995b-2e09e35ac0fb">



## PR checklist: submitters


## PR checklist: reviewers


The larger the PR, the stricter we should be about these points.
